File: frontend/src/api/services/matchService.ts
import api from '../axios';
import type { MatchEvent, MatchInvitation, MatchResult, MatchStatus, Visibility } from '../../types';

export interface CreateMatchRequest {
    hostTeamId: number;
    matchDate: string;
    startTime: string;
    endTime?: string;
    fieldId?: number;
    description?: string;
    visibility?: Visibility;
}

export interface UpdateMatchRequest extends Partial<CreateMatchRequest> {
    matchId: number;
    opponentTeamId?: number;
    status?: MatchStatus;
}

export interface MatchSearchParams {
    teamId?: number;
    status?: MatchStatus;
    fromDate?: string;
    toDate?: string;
    page?: number;
    limit?: number;
}

export interface SendInvitationRequest {
    matchId: number;
    invitedTeamId: number;
    message?: string;
}

export interface PaginatedResponse<T> {
    data: T[];
    total: number;
    page: number;
    limit: number;
    totalPages: number;
}

// Match service functions - Real API calls
export const matchService = {
    /**
     * Create a new match event
     */
    createMatch: async (data: CreateMatchRequest): Promise<MatchEvent> => {
        const response = await api.post<MatchEvent>('/matches', data);
        return response.data;
    },

    /**
     * Get match by ID
     */
    getMatchById: async (matchId: number): Promise<MatchEvent> => {
        const response = await api.get<MatchEvent>(`/matches/${matchId}`);
        return response.data;
    },

    /**
     * Update match
     */
    updateMatch: async ({ matchId, ...data }: UpdateMatchRequest): Promise<MatchEvent> => {
        const response = await api.put<MatchEvent>(`/matches/${matchId}`, data);
        return response.data;
    },

    /**
     * Cancel match
     */
    cancelMatch: async (matchId: number, _reason?: string): Promise<MatchEvent> => {
        const response = await api.put<MatchEvent>(`/matches/${matchId}/cancel`);
        return response.data;
    },

    /**
     * Get matches for a team
     */
    getTeamMatches: async (teamId: number, params?: MatchSearchParams): Promise<PaginatedResponse<MatchEvent>> => {
        const response = await api.get<MatchEvent[]>(`/matches/team/${teamId}`, { params });
        return {
            data: response.data,
            total: response.data.length,
            page: params?.page || 1,
            limit: params?.limit || 20,
            totalPages: 1,
        };
    },

    /**
     * Get player's match schedule (by user_id, looks up player first)
     */
    getPlayerSchedule: async (userId: number): Promise<MatchEvent[]> => {
        const response = await api.get<MatchEvent[]>(`/players/user/${userId}/schedule`);
        return response.data;
    },

    /**
     * Send match invitation
     */
    sendInvitation: async (data: SendInvitationRequest): Promise<MatchInvitation> => {
        const response = await api.post<MatchInvitation>(`/matches/${data.matchId}/invitations`, {
            invitedTeamId: data.invitedTeamId,
            message: data.message,
        });
        return response.data;
    },

    /**
     * Get pending invitations for a team
     */
    getPendingInvitations: async (teamId: number): Promise<MatchInvitation[]> => {
        const response = await api.get<MatchInvitation[]>(`/teams/${teamId}/match-invitations`);
        return response.data;
    },

    /**
     * Respond to invitation
     */
    respondInvitation: async (invitationId: number, accept: boolean): Promise<MatchInvitation> => {
        const action = accept ? 'accept' : 'decline';
        const response = await api.put<MatchInvitation>(`/matches/invitations/${invitationId}/${action}`);
        return response.data;
    },

    /**
     * Record match result
     */
    recordResult: async (matchId: number, homeScore: number, awayScore: number, notes?: string): Promise<MatchResult> => {
        const response = await api.post<MatchResult>(`/matches/${matchId}/result`, {
            homeScore,
            awayScore,
            notes,
        });
        return response.data;
    },

    /**
     * Get match result
     */
    getMatchResult: async (matchId: number): Promise<MatchResult | null> => {
        try {
            const response = await api.get<MatchResult>(`/matches/${matchId}/result`);
            return response.data;
        } catch {
            return null;
        }
    },
};

export default matchService;

File: backend/app/services/match_service.py
"""
MatchService - Match event business logic.
Maps to MatchController in class diagram.
"""
from typing import List, Optional
from datetime import datetime, date, time
from sqlalchemy.ext.asyncio import AsyncSession

from app.repositories.match_repository import MatchRepository, InvitationRepository, AttendanceRepository
from app.repositories.team_repository import TeamRepository
from app.services.notification_service import NotificationService
from app.models.match import MatchEvent, MatchInvitation, AttendanceRecord, MatchResult
from app.models.enums import MatchStatus, Visibility, InvitationStatus, NotificationType

class MatchService:
    """Service handling match business logic."""
    
    def __init__(self, db: AsyncSession):
        self.db = db
        self.match_repo = MatchRepository(db)
        self.invitation_repo = InvitationRepository(db)
        self.attendance_repo = AttendanceRepository(db)
        self.team_repo = TeamRepository(db)
        self.notification_service = NotificationService(db)
    
    async def create_match(
        self,
        host_team_id: int,
        match_date: date,
        start_time: time,
        end_time: time = None,
        field_id: int = None,
        visibility: Visibility = Visibility.PUBLIC,
        description: str = None,
    ) -> MatchEvent:
        """Create a new match event."""
        match = MatchEvent(
            host_team_id=host_team_id,
            match_date=match_date,
            start_time=start_time,
            end_time=end_time,
            field_id=field_id,
            visibility=visibility,
            description=description,
            status=MatchStatus.LOOKING_FOR_FIELD if not field_id else MatchStatus.SCHEDULED,
        )
        await self.match_repo.save(match)
        await self.match_repo.commit()
        return match
    
    async def get_match_by_id(self, match_id: int) -> Optional[MatchEvent]:
        """Get match by ID."""
        return await self.match_repo.find_by_id(match_id)
    
    async def get_matches_by_team(self, team_id: int) -> List[MatchEvent]:
        """Get matches for a team."""
        return await self.match_repo.find_by_team(team_id)
    
    async def update_match(self, match: MatchEvent, **kwargs) -> MatchEvent:
        """Update match event."""
        for key, value in kwargs.items():
            if hasattr(match, key) and value is not None:
                setattr(match, key, value)
        
        await self.match_repo.update(match)
        await self.match_repo.commit()
        return match
    
    async def cancel_match(self, match: MatchEvent) -> bool:
        """Cancel a match."""
        match.status = MatchStatus.CANCELLED
        await self.match_repo.update(match)
        await self.match_repo.commit()
        return True
    
    # --- Invitations ---
    
    async def send_invitation(
        self,
        match: MatchEvent,
        invited_team_id: int,
        message: str = None
    ) -> MatchInvitation:
        """Send match invitation."""
        invitation = MatchInvitation(
            match_id=match.match_id,
            inviting_team_id=match.host_team_id,
            invited_team_id=invited_team_id,
            status=InvitationStatus.PENDING,
            message=message,
        )
        await self.invitation_repo.save(invitation)
        
        # Notify invited team leader
        invited_team = await self.team_repo.find_by_id(invited_team_id)
        if invited_team:
            await self.notification_service.create_notification(
                user_id=invited_team.leader_id,
                notification_type=NotificationType.MATCH_INVITE,
                title="Match Invitation",
                message=f"You have been invited to a match.",
                related_entity_id=match.match_id,
                related_entity_type="Match"
            )
            
        await self.invitation_repo.commit()
        return invitation
    
    async def respond_to_invitation(
        self,
        invitation: MatchInvitation,
        accept: bool
    ) -> bool:
        """Accept or decline match invitation."""
        invitation.responded_at = datetime.utcnow()
        
        if accept:
            invitation.status = InvitationStatus.ACCEPTED
            # Update match with opponent
            match = await self.match_repo.find_by_id(invitation.match_id)
            if match:
                match.opponent_team_id = invitation.invited_team_id
                if match.status == MatchStatus.LOOKING_FOR_FIELD:
                    match.status = MatchStatus.SCHEDULED
                await self.match_repo.update(match)
        else:
            invitation.status = InvitationStatus.DECLINED
        
        await self.invitation_repo.update(invitation)
        await self.invitation_repo.commit()
        return True
    
    async def get_match_attendance(self, match_id: int) -> List[AttendanceRecord]:
        """Get attendance records for a match."""
        return await self.attendance_repo.find_by_match(match_id)
